src/schema/sources.ts

```diff
@@ -347,56 +374,146 @@ const sourceToGQL = (source: Source): GQLSource => ({
 });
 
 export enum SourcePermissions {
-  View,
-  Post,
-  Leave,
-  Delete,
-  Edit,
+  View = 'view',
+  Post = 'post',
+  PostLimit = 'post_limit',
+  PostDelete = 'post_delete',
+  MemberRemove = 'member_remove',
+  ModeratorAdd = 'moderator_add',
+  ModeratorRemove = 'moderator_remove',
+  InviteDisable = 'invite_disable',
+  Leave = 'leave',
+  Delete = 'delete',
+  Edit = 'edit',
 }
 
+export const roleSourcePermissions: Record<
+  SourceMemberRoles,
+  SourcePermissions[]
+> = {
+  owner: [
+    SourcePermissions.View,
+    SourcePermissions.Post,
+    SourcePermissions.PostLimit,
+    SourcePermissions.PostDelete,
+    SourcePermissions.MemberRemove,
+    SourcePermissions.ModeratorAdd,
+    SourcePermissions.ModeratorRemove,
+    SourcePermissions.InviteDisable,
+    SourcePermissions.Edit,
+    SourcePermissions.Delete,
+  ],
+  moderator: [
+    SourcePermissions.View,
+    SourcePermissions.Post,
+    SourcePermissions.PostDelete,
+    SourcePermissions.MemberRemove,
+    SourcePermissions.Edit,
+    SourcePermissions.Leave,
+  ],
+  member: [
+    SourcePermissions.View,
+    SourcePermissions.Post,
+    SourcePermissions.Leave,
+  ],
+};
+
+const requireGreaterAccessPrivilege: Partial<
+  Record<SourcePermissions, boolean>
+> = {
+  [SourcePermissions.MemberRemove]: true,
+};
+
+type BaseSourceMember = Pick<SourceMember, 'role'>;
+
+export const hasGreaterAccessCheck = (
+  loggedUser: BaseSourceMember,
+  member: BaseSourceMember,
+) => {
+  const memberRank = roleRank[member.role];
+  const loggedUserRank = roleRank[loggedUser.role];
+  const hasGreaterAccess = loggedUserRank > memberRank;
+
+  if (!hasGreaterAccess) {
+    throw new ForbiddenError('Access denied!');
+  }
+};
+
+interface SourcePermissionsProps {
+  permission: SourcePermissions;
+  validateRankAgainstId: string;
+}
+
+const hasPermissionCheck = (
+  source: Source | GQLSource,
+  member: BaseSourceMember,
+  permission: SourcePermissions,
+  validateRankAgainst?: BaseSourceMember,
+) => {
+  if (
+    source.type !== SourceType.Squad &&
+    [SourcePermissions.Post, SourcePermissions.Leave].includes(permission)
+  ) {
+    return false;
+  }
+
+  if (validateRankAgainst) {
+    hasGreaterAccessCheck(member, validateRankAgainst);
+  }
+
+  const rolePermissions =
+    roleSourcePermissions[member.role] ?? roleSourcePermissions.member;
+
+  return rolePermissions.includes(permission);
+};
+
+export const getUserPermissions = (
```

Comment:
What does this function do exactly? If I understand correctly you can just return the right array of permissions based on the role

Reply:
I think I forgot to remove this when I migrated to the new process. My bad. Will clear this one out.